### ISSUE TITLE
Headless/CLI debugger

### ISSUE THREAD

**Issue by ASzc** [stella-emu/stella#509]
With PromptWidget being able to control almost everything to do with the debugger, Stella is really close to being able to run as a CLI, or from a script. If this were possible, it would enable automatic test suites for homebrew games, and tool assisted speedruns (TAS).

Autoexec scripts aren't quite enough, as there's no way to get information out of the debugger (`saveses` doesn't work for autoexec history), and it loads the GUI. Disabling `GUI_SUPPORT` is also non-viable, since it completely disables the GUI at compile time, and also disables the debugger.

I've had a look, and it seems the PromptWidget is pretty tightly coupled with the GUI and shell management code. In [my branch](https://github.com/ASzc/stella/tree/headless) I've done some initial work towards a CLI-only mode, but it's crude, and doesn't work all that well.

Ideally I'd like the mode to work like this:

- Mode enabled at runtime from command line arguments
- No GUI is loaded at any point
- Sound is disabled
- Full emulation of video is rendered, but not displayed to the user 
- Current frame can be dumped to a file by command, with full path specifiable as an argument
- Commands accepted from stdin, output returned on stdout from the commands (a REPL)
- Ability to query the specific values out of RAM via symbol alias
  - Optionally, have assertions built in as a command, but it's possible to implement these externally with a proper query feature.

**Comment by thrust26** [comment]
@ASzc So what are your plans here? Are you going forward with this? 

I suppose for the Stella team this will have low priority and most likely may never happen.

**Comment by sa666666** [comment]
I agree that it would nice to have, but also with @thrust26 that I don't see any time in the immediate future where I can even consider it.

**Comment by ASzc** [comment]
Priority situation makes sense. I'm not the most skilled C++ programmer, but I may be able to complete this myself.

It'd help to have some guidance around how the maintainers would prefer such a feature to be added. For example:

- I've just stuck my own thread onto stdio, whereas there seems to be an the event loop that'd normally handle input in Stella. Should I just cut out the event loop entirely, or try to integrate into that more? Is the event loop necessary for emulation to occur?

- The user experience of the command line REPL can't benefit from the UI code in PromptWidget, it's not really transferable to a terminal environment. I could leave the CLI REPL minimal as it is, or bring in a dependency on something like [cpp-linenoise](https://github.com/yhirose/cpp-linenoise/blob/master/linenoise.hpp) to allow line editing and recall. Is a new dependency acceptable? Is a 2-clause BSD library acceptable? How would one add the dependency into Stella's build process?

**Comment by sa666666** [comment]
Just wondering if any more work has been done on this?  If not, I guess we can close this issue.
